### DISCUSSION TITLE
OpenSSL 3.2 Alpha - CLI for QUIC?

### DISCUSSION THREAD

**Discussion by ckcr4lyf** [openssl/openssl#22072] [Q&A] [answered]
Hey guys. I like to use the OpenSSL CLI a lot for triaging TLS connections, as well as wrangling keys & certs.

I saw the announcement for QUIC support in OpenSSL 3.2 Alpha, I was wondering if I could use it via the CLI?

I gave https://www.openssl.org/docs/manmaster/man7/openssl-quic.html a read, but it seems it is only exposed via the library.

Cheers.

**Comment by paulidale** [comment, marked as answer]
There is a `-quic` option to [`s_client`](https://www.openssl.org/docs/manmaster/man1/openssl-s_client.html).

**Reply by mattcaswell** [reply]
Note that ALPN is mandatory in QUIC so you must also supply the `-alpn` option to s_client.

**Reply by mattcaswell** [reply]
Also note that so far s_client only supports a single stream (although the underlying library does support multiple streams)

**Reply by t8m** [reply]
To make it clear - you should be able to connect with HTTP/3.0 servers if you supply the right ALPN but you won't be able to communicate further with such server because that would require HTTP/3.0 support in s_client.

**Reply by beldmit** [reply]
Doesn't it mean we should either implicitly add it or explicitly complain (have not seen the code yet)

**Reply by mattcaswell** [reply]
We do complain:

````
$ openssl s_client -quic
s_client: QUIC requires ALPN to be specified (e.g. "h3" for HTTP/3) via the -alpn option
````

**Reply by ckcr4lyf** [reply]
Thanks, this works perfectly. I tried:

```
$ ./openssl s_client -connect cloudflare-quic.com:443 -quic -alpn "h3"
```

And it seems to have successfully completed a handshake. Wireshark also seems legit:

![image](https://github.com/openssl/openssl/assets/6680615/72dfd58f-d329-4a7f-9482-922f031c36d0)

Though as t8m mentioned, I can't actually try and make an HTTP request or similar (at least on the CLI) since `s_client` doesn't support HTTP/3.0 (if I understand correctly)

**Reply by mattcaswell** [reply]
You can also add the "-trace" option to get additional tracing output from s_client (e.g. to see the handshake).

**Reply by ckcr4lyf** [reply]
That's pretty cool, very detailed info. Thanks.

Out of curiosity - does OpenSSL implement QUIC from scratch on top of UDP sockets provided by the kernel? Or does it use any existing libraries? I saw `quiche` in the top level as a submodule but some casual searching seems to indicate it is used just for testing.

**Reply by mattcaswell** [reply]
It is a from-scratch implementation.

**Comment by beldmit** [comment]
> Note that ALPN is mandatory in QUIC so you must also supply the `-alpn` option to s_client.
